sympy/functions/combinatorial/factorials.py

```diff
@@ -868,192 +920,440 @@ def fdiff(self, argindex=1):
             raise ArgumentIndexError(self, argindex)
 
     @classmethod
-    def _eval(self, n, k):
+    def _eval(cls, n, k):
         # n.is_Number and k.is_Integer and k != 1 and n != k
-
-        if k.is_Integer:
-            if n.is_Integer and n >= 0:
-                n, k = int(n), int(k)
-
-                if k > n:
-                    return S.Zero
-                elif k > n // 2:
-                    k = n - k
-
-                if HAS_GMPY:
-                    from sympy.core.compatibility import gmpy
-                    return Integer(gmpy.bincoef(n, k))
-
-                d, result = n - k, 1
-                for i in range(1, k + 1):
-                    d += 1
-                    result = result * d // i
-                return Integer(result)
-            else:
-                d, result = n - k, 1
-                for i in range(1, k + 1):
-                    d += 1
-                    result *= d
-                    result /= i
-                return result
+        # and k > 0 (and, if n is positive, k < n - k)
+        Ntype = Float if k.has(Float) or n.has(Float) else Integer
+        k = as_int(k, strict=False)
+        direct = n.is_nonnegative
+        if direct:
+            try:
+                n = as_int(n, strict=False)
+            except ValueError:
+                direct = False
+        if direct:
+
+            if HAS_GMPY:
+                from sympy.core.compatibility import gmpy
+                return Ntype(gmpy.bincoef(n, k))
+
+            d, result = n - k, 1
+            for i in range(1, k + 1):
+                d += 1
+                result = result * d // i
+            return Ntype(result)
+        else:
+            return Mul(*[n + i for i in range(1 - k, 1)])/factorial(k)
 
     @classmethod
     def eval(cls, n, k):
+        from sympy import Min
         n, k = map(sympify, (n, k))
-        d = n - k
-        n_nonneg, n_isint = n.is_nonnegative, n.is_integer
-        if k.is_zero or ((n_nonneg or n_isint is False)
-                and d.is_zero):
-            return S.One
-        if (k - 1).is_zero or ((n_nonneg or n_isint is False)
-                and (d - 1).is_zero):
-            return n
-        if k.is_integer:
-            if k.is_negative or (n_nonneg and n_isint and d.is_negative):
+        if n.is_zero:   #
+            n = S.Zero  # XXX is this anti-SymPy?
+        if k.is_zero:   #
+            k = S.Zero  #
```

Comment:
Why is this needed?

Reply:
Then any code that uses `is S.Zero` will work without modification.

Reply:
I think it is better to fix the places that check `is S.Zero`. Did some come up during this PR?

Reply:
The code has been removed. Let's see how tests go.